distributed/shuffle/_buffer.py

```diff
@@ -0,0 +1,231 @@
+import weakref
+from collections import defaultdict
+from collections.abc import Iterator
+from typing import TYPE_CHECKING, Any, Generic, Sized, TypeVar
+
+from tornado.ioloop import IOLoop
+
+from distributed.metrics import time
+from distributed.shuffle._limiter import ResourceLimiter
+
+if TYPE_CHECKING:
+    import pyarrow as pa
+
+logger = logging.getLogger("distributed.shuffle")
+
+ShardType = TypeVar("ShardType", bound=Sized)
+
+
+class ShardsBuffer(Generic[ShardType]):
+
+    shards: defaultdict[str, list[ShardType]]
+    sizes: defaultdict[str, int]
+    _exception: None | Exception
+
+    _queues: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
+
+    @property
+    def _instances(self) -> set:
+        raise NotImplementedError()
+
+    def __init__(
+        self,
+        memory_limiter: ResourceLimiter | None,
+        concurrency_limit: int = 2,
+        max_message_size: int = -1,
+    ) -> None:
+        self._closed = False
+        self.shards = defaultdict(list)
+        self.sizes = defaultdict(int)
+        self._exception = None
+        self.concurrency_limit = concurrency_limit
+        self._done = False
+        self.memory_limiter = memory_limiter
+        self.diagnostics: dict[str, float] = defaultdict(float)
+        self._tasks = [
+            asyncio.create_task(self.worker()) for _ in range(concurrency_limit)
+        ]
+        self._shards_available = asyncio.Condition()
+        self._flush_lock = asyncio.Lock()
+        self.max_message_size = max_message_size
+
+        self.bytes_total = 0
+        self.bytes_memory = 0
+        self.bytes_written = 0
+        self.bytes_read = 0
+
+    def heartbeat(self) -> dict[str, Any]:
+        return {
+            "memory": self.bytes_memory,
+            "total": self.bytes_total,
+            "buckets": len(self.shards),
+            "written": self.bytes_written,
+            "read": self.bytes_read,
+            "diagnostics": self.diagnostics,
+            "memory_limit": self.memory_limiter._maxvalue if self.memory_limiter else 0,
+        }
+
+    async def process(self, id: str, shards: list[pa.Table], size: int) -> None:
+        try:
+            start = time()
+            try:
+                await self._process(id, shards)
+                self.bytes_written += size
+
+            except Exception as e:
+                self._exception = e
+                self._done = True
+            stop = time()
+
+            self.diagnostics["avg_size"] = (
+                0.98 * self.diagnostics["avg_size"] + 0.02 * size
+            )
+            self.diagnostics["avg_duration"] = 0.98 * self.diagnostics[
+                "avg_duration"
+            ] + 0.02 * (stop - start)
+        finally:
+            if self.memory_limiter:
+                await self.memory_limiter.release(size)
+            self.bytes_memory -= size
+
+    async def _process(self, id: str, shards: list[ShardType]) -> None:
+        raise NotImplementedError()
```

Comment:
What would be the benefit of this?

Reply:
Marking `ShardsBuffer` as an `abc.ABC` as well as decorating `_process` with `@abc.abstractmethod` feels mildly cleaner since `ShardsBuffer` should not be initialized and any concrete subclass needs to implement `_process`[1]. This would allow `mypy` to catch errors such as trying to instantiate `ShardsBuffer` or a subclass that did not implement `_process`. Similarly, Python would throw somewhat more informative errors at runtime.

I guess this is mainly a question about conventions: If a class is an abstract class, should it always be marked as such or do we not care about that. If it is the latter, why do we not care? (We don't see any value in this case is a perfectly valid answer).

Feel free to shut this discussion down, the case at hand revolves around an abstract class that will have two concrete subclasses for the foreseeable future and is in a private module, so I'm rather interested in whether we have some conventions/guidelines around marking ABCs, not splitting hairs about this specific case.

[1] `read` not so much, which leads me off-trail to asking whether `read` belongs here in the class hierarchy, but that's a fringe problem that requires more changes and I don't want to deal with this right now.